components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,416 @@
+
+  const auto keyring_info = keyring_service_->GetKeyringInfoSync(
+      brave_wallet::mojom::kDefaultKeyringId);
+  std::vector<std::string> account_addresses;
+  for (const auto& account_info : keyring_info->account_infos) {
+    account_addresses.push_back(account_info->address);
+  }
+
+  if (account_addresses.empty()) {
+    OnDiscoverEthAllowancesCompleted({});
+    return;
+  }
+
+  const auto token_list_map =
+      BlockchainRegistry::GetInstance()->GetEthTokenListMap(
+          GetChainIdsForAlowanceDiscovering());
+
+  base::flat_map<std::string, base::Value::List> chain_id_to_contract_addresses;
+
+  for (const auto& [chain_id, token_list] : token_list_map) {
+    for (const auto& token : token_list) {
+      chain_id_to_contract_addresses[chain_id].Append(token->contract_address);
+    }
+  }
+
+  if (chain_id_to_contract_addresses.empty()) {
+    OnDiscoverEthAllowancesCompleted({});
+    return;
+  }
+
+  const auto& allowance_cache_dict =
+      prefs_->GetDict(kBraveWalletEthAllowancesCache);
+
+  if (!tasks.empty()) {
+    tasks.clear();
+  }
+
+  int task_id(0);
+  const auto approval_topic_hash = KeccakHash(kApprovalTopicHash);
+  for (const auto& account_address : account_addresses) {
+    for (const auto& [chain_id, contract_addresses] :
+         chain_id_to_contract_addresses) {
+      const auto* chain_cached_data = allowance_cache_dict.FindDict(chain_id);
+
+      const auto* last_block_number_ptr =
+          chain_cached_data ? chain_cached_data->FindString(kLastBlockNumber)
+                            : nullptr;
+      const std::string last_block_number =
+          last_block_number_ptr != nullptr ? *last_block_number_ptr : "";
+
+      std::string hex_account_address;
+      if (!PadHexEncodedParameter(account_address, &hex_account_address)) {
+        continue;
+      }
+      auto internal_callback = base::BindOnce(
+          &EthAllowanceManager::OnGetAllowances, weak_ptr_factory_.GetWeakPtr(),
+          task_id, chain_id, hex_account_address);
+      tasks.insert_or_assign(task_id,
+                             std::make_unique<EthAllowanceTask>(task_id));
+
+      base::Value::List topics;
+      topics.Append(approval_topic_hash);
+      topics.Append(std::move(hex_account_address));
+
+      base::Value::Dict filter_options;
+      filter_options.Set(kAddress, contract_addresses.Clone());
+      filter_options.Set(kTopics, std::move(topics));
+      uint256_t block_number{0};
+      if (!last_block_number.empty() &&
+          HexValueToUint256(last_block_number, &block_number)) {
+        filter_options.Set(kFromBlock, Uint256ValueToHex(++block_number));
+      } else {
+        filter_options.Set(kFromBlock, kEarliestBlock);
+      }
+      filter_options.Set(kToBlock, kLatestBlock);
+      json_rpc_service_->EthGetLogs(chain_id, std::move(filter_options),
+                                    std::move(internal_callback));
+      task_id++;
+    }
+  }
+}
+
+void EthAllowanceManager::Reset() {
```

Comment:
Should call
`weak_ptr_factory_.InvalidateWeakPtrs();`
so any potential callbacks sent to json_rpc_service_ will not return back.